### DISCUSSION TITLE
Divergent autosubmit functionality in Safari

### DISCUSSION THREAD

**Discussion by jconley88** [unpoly/unpoly#471] [Q&A]
I'm not sure if this is a bug or if there's something external causing this behavior.

I have a page in a rails project, that is loaded via a POST and then a redirect to a GET page. On this final GET page, I have an unpoly autosubmit form which works properly in Firefox and Chrome. However, in Safari, every time a form input is changed, the entire page is refreshed and the input loses focus. It is a textarea input designed for a long form response and is essentially non functional. I've tracked the issue to a difference in how unpoly decides the page was loaded.

In Firefox, after the POST->redirect->GET, the [initialRequestMethod](https://github.com/unpoly/unpoly/blob/2.x-stable/src/unpoly/protocol.js#L611) is set to "GET". This means that [canPushState](https://github.com/unpoly/unpoly/blob/3731facecc7a6be26abaa9c27a522f46009ab6ff/src/unpoly/browser.js#L38) is true and autosubmit will work seamlessly without interrupting the user typing in the textarea.

In Safari, after the POST->redirect->GET, the initialRequestMethod gets set and memoized to "POST". This causes "canPushState" to be false, which triggers a full refresh on submission of the form. This form is submitted via a PATCH (POST), returns a redirect to a GET and sets the page up for another full refresh as initialRequestMethod is now "PATCH".

Both browser request and response headers appear to be identical. The *_up_method* cookie is initially set to POST and then unset via expiration on the the subsequent redirected GET. I believe that this is handled via the unpoly-rails gem. It seems to me that Firefox is unsetting the cookie before unpoly is initialized and Safari is doing it after.

Safari is a common browser and a POST->redirect->GET is a common scenario. I haven't seen anyone else with the same issue which makes me wonder if my code is doing something unusual that I don't know about. I've tried multiple versions of unpoly and the unpoly-rails gem, including 2.6.1 and 2.7.2. Safari version: 16.4.

Has anyone run into this? Is there an obvious setting or workaround that I may have missed? I'm currently considering reimplementing the autosubmit to not rely on *up.submit()*. It's not a great solution as there is other unpoly functionality on that page which I'd also have to implement a workaround for.

**Comment by triskweline** [comment]
We haven't seen this behavior in our own Rails apps.

To look into this further, I would need to make a minimal Rails app to reproduce this issue. Is this something you could help with?

Meanwhile a workaround could be forcing `canPushState()` to return `true` by including this snippet after requiring Unpoly:

```js
up.browser.canPushState = () => true
```

**Comment by jconley88** [comment]
Thanks for your response! I appreciate knowing that you haven't seen this. The workaround works great for now. I'll see if I can generate a minimal app that reproduces the issue.

**Comment by jconley88** [comment]
@triskweline I have a minimal [rails app](https://github.com/jconley88/unpoly_autosubmit) up on github. It should be pretty self explanatory based on the README and instructions on pages. The issue is only reproducible in a normal Safari window as far as I can tell.

In a private Safari window autosubmit operates properly. I normally use this to rule out local environment issues, but to my knowledge I have not modified anything about Safari. I also had a colleague confirm the issue.
